vault/resource_plugin.go

```diff
@@ -0,0 +1,270 @@
+				Optional:    true,
+			},
+			fieldRuntime: {
+				Type:        schema.TypeString,
+				Description: "Vault plugin runtime to use if oci_image is specified.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func pluginWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	name := d.Get(consts.FieldName).(string)
+	version := d.Get(consts.FieldVersion).(string)
+	id := idFromPlugin(pluginType.String(), name, version)
+
+	if diagErr := versionedPluginsSupported(meta, version); diagErr != nil {
+		return diagErr
+	}
+
+	ociImage := d.Get(fieldOCIImage).(string)
+	runtime := d.Get(fieldRuntime).(string)
+	if diagErr := containerizedPluginsSupported(meta, ociImage, runtime); diagErr != nil {
+		return diagErr
+	}
+
+	log.Printf("[DEBUG] Writing plugin %q", id)
+	err = client.Sys().RegisterPluginWithContext(ctx, &api.RegisterPluginInput{
+		Type:     pluginType,
+		Name:     name,
+		Version:  version,
+		SHA256:   d.Get(fieldSHA256).(string),
+		Command:  d.Get(fieldCommand).(string),
+		Args:     util.ToStringArray(d.Get(fieldArgs).([]interface{})),
+		Env:      util.ToStringArray(d.Get(fieldEnv).([]interface{})),
+		OCIImage: ociImage,
+		Runtime:  runtime,
+	})
+	if err != nil {
+		return diag.Errorf("error updating plugin %q: %s", id, err)
+	}
+	log.Printf("[DEBUG] Wrote plugin %q", id)
+
+	d.SetId(id)
+
+	return pluginRead(ctx, d, meta)
+}
+
+func pluginRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	typ, name, version := pluginFromID(d.Id())
+	if typ == "" || name == "" {
+		diag.Errorf("invalid ID %q, must be of form :type/name/:name or :type/version/:version/name/:name", d.Id())
+	}
+
+	if diagErr := versionedPluginsSupported(meta, version); diagErr != nil {
+		return diagErr
+	}
+
+	pluginType, err := api.ParsePluginType(typ)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := client.Sys().GetPluginWithContext(ctx, &api.GetPluginInput{
+		Type:    pluginType,
+		Name:    name,
+		Version: version,
+	})
+
+	if err != nil && util.Is404(err) {
+		log.Printf("[WARN] plugin %q not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	} else if err != nil {
+		return diag.Errorf("error reading plugin %q: %s", d.Id(), err)
+	}
+
+	result := map[string]any{
+		consts.FieldType:    typ,
+		consts.FieldName:    name,
+		consts.FieldVersion: version,
+		fieldSHA256:         resp.SHA256,
+		fieldCommand:        resp.Command,
+		fieldOCIImage:       resp.OCIImage,
```

Comment:
might be a nit:
is it worth checking is containerized plugins are supported before attempting to set the value to the TF state? In the usual case if Vault returns OCI Image info, it would mean that containerized plugins are available, so we should be okay to trust this value. I noticed however we were checking `versionedPluginsSupported` above, so was curious if there could be any edge cases where checking the Vault version before setting to state might be worth it

Reply:
My reasoning here is that anything could be user-supplied input in the write function, so for any feature that's somewhat new we should validate it's supported in Vault.

However, in the read function, only the elements we pull from the ID _may_ be user-supplied, everything else _must_ come from Vault. So in the read function I've validated the version because it's in the ID, but I don't see any edge case where containerised plugins settings could be set but they're not supported, because they must have come from Vault.

Does that sound correct to you?